### PR TITLE
[rllib] Refactor pytorch custom model support

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

- Clean up the pytorch model API to support RNNs, Dict / Tuple spaces
- Unify QMIX RNN model with model catalog

I expect we'll have to make more changes (and add more tests) as we implement PyTorch support more fully; this is just an initial cleanup to better unify QMIX with pytorch A3C.

https://github.com/ray-project/ray/issues/3365

### CODE REVIEW

File: python/ray/rllib/agents/qmix/qmix_policy_graph.py [hunk 1]
```diff
@@ -292,8 +301,8 @@ def to_batches(arr):
     @override(PolicyGraph)
     def get_initial_state(self):
         return [
-            self.model.init_hidden().numpy().squeeze()
-            for _ in range(self.n_agents)
+            s.expand([self.n_agents, -1]).numpy()
```
Comment (ericl) [inline python/ray/rllib/agents/qmix/qmix_policy_graph.py]: The main change here is that we return a list of `[num_agents, h_size_i] for i in n_state_tensors`, rather than `[h_size] for i in n_agents` previously (which was kind of unnatural and didn't generalize well to multiple state tensors).

Relatedly, I'm regretting allowing multiple hidden state tensors, we probably should have just required them to be fused into one element.

File: python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ def __init__(self, obs_space, action_space, config):
 
     @override(TorchPolicyGraph)
     def extra_action_out(self, model_out):
-        return {"vf_preds": var_to_np(model_out[1])}
+        return {"vf_preds": var_to_np(model_out[2])}
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py]:
```
def var_to_np(var):
    return var.cpu().detach().numpy()
```

It looks like that does the cpu() and detach() as well, do we still need that?

[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py]: I think probably not; all this pytorch API stuff changed a lot since.

File: python/ray/rllib/models/pytorch/fcnet.py [hunk 3]
```diff
@@ -3,30 +3,27 @@
...
-            activation = nn.Tanh
-        elif fcnet_activation == "relu":
-            activation = nn.ReLU
-        logger.info("Constructing fcnet {} {}".format(hiddens, activation))
+class FullyConnectedNetwork(TorchModel):
+    """Generic fully connected network."""
 
+    def __init__(self, obs_space, num_outputs, options):
+        TorchModel.__init__(self, obs_space, num_outputs, options)
+        hiddens = options.get("fcnet_hiddens")
```
Comment (richardliaw) [inline python/ray/rllib/models/pytorch/fcnet.py]: should this have a default?

Comment (richardliaw) [reply python/ray/rllib/models/pytorch/fcnet.py]: right now, if not provided, this fails at line 26 (which you may as well make this `options["fcnet_hiddens"]`

File: python/ray/rllib/models/pytorch/visionnet.py [hunk 4]
```diff
@@ -39,31 +36,20 @@ def _build_layers(self, inputs, num_outputs, options):
...
+        features = self._hidden_layers(input_dict["obs"])
+        logits = self._logits(features)
+        value = self._value_branch(features).squeeze(1)
+        return logits, features, value, hidden_state
 
-        args:
-            obs: observations and features"""
-        res = self._convs(obs)
+    def _hidden_layers(self, obs):
+        res = self._convs(obs.permute(0, 3, 1, 2))
```
Comment (richardliaw) [inline python/ray/rllib/models/pytorch/visionnet.py]: can this be documented somewhere?

File: python/ray/rllib/agents/qmix/qmix_policy_graph.py [hunk 5]
```diff
@@ -219,11 +228,10 @@ def compute_actions(self,
...
             actions = (pick_random * random_actions +
                        (1 - pick_random) * masked_q_values.max(dim=2)[1])
-            actions = var_to_np(actions)
-            hiddens = var_to_np(hiddens)
+            actions = actions.numpy()
+            hiddens = [s.numpy() for s in hiddens]
 
-        return (TupleActions(list(actions.transpose([1, 0]))),
-                hiddens.transpose([1, 0, 2]), {})
+        return TupleActions(list(actions.transpose([1, 0]))), hiddens, {}
```
Comment (richardliaw) [inline python/ray/rllib/agents/qmix/qmix_policy_graph.py]: is the second `[1, 0, 2]` not needed anymore?

**Review by richardliaw — APPROVED** [review]

File: python/ray/rllib/models/pytorch/fcnet.py (hunk 3 above)
Comment (ericl) [reply python/ray/rllib/models/pytorch/fcnet.py]: It's filled in by agent, so it should be ok either way.

File: python/ray/rllib/models/pytorch/visionnet.py (hunk 4 above)
Comment (ericl) [reply python/ray/rllib/models/pytorch/visionnet.py]: Done

File: python/ray/rllib/agents/qmix/qmix_policy_graph.py (hunk 5 above)
Comment (ericl) [reply python/ray/rllib/agents/qmix/qmix_policy_graph.py]: Yeah (see comment above), hiddens are handled differently now.
